src/core.jl

```diff
@@ -0,0 +1,72 @@
+# Base Type
+## API:
+### initialparameters(rng, l) --> Must return a NamedTuple/ComponentArray
+### initialstates(rng, l)     --> Must return a NamedTuple
+### parameterlength(l)        --> Integer
+### statelength(l)            --> Integer
+### l(x, ps, st)
+
+abstract type AbstractExplicitLayer end
+
+initialparameters(::AbstractRNG, ::Any) = ComponentArray{Float32}()
+initialparameters(l) = initialparameters(Random.GLOBAL_RNG, l)
+initialparameters(rng::AbstractRNG, l::NamedTuple{fields}) where {fields} =
+    ComponentArray(NamedTuple{fields}(initialparameters.(rng, values(l))))
+
+initialstates(::AbstractRNG, ::Any) = NamedTuple()
+initialstates(l) = initialstates(Random.GLOBAL_RNG, l)
+initialstates(rng::AbstractRNG, l::NamedTuple{fields}) where {fields} =
+    NamedTuple{fields}(initialstates.(rng, values(l)))
+
+parameterlength(l::AbstractExplicitLayer) = parameterlength(initialparameters(l))
+parameterlength(nt::Union{NamedTuple,Tuple}) = length(nt) == 0 ? 0 : sum(parameterlength, nt)
+parameterlength(a::AbstractArray) = length(a)
+parameterlength(x) = 0
+
+statelength(l::AbstractExplicitLayer) = statelength(initialstates(l))
+statelength(nt::Union{NamedTuple,Tuple}) = length(nt) == 0 ? 0 : sum(statelength, nt)
+statelength(a::AbstractArray) = length(a)
+statelength(x::Union{Number,Symbol}) = 1
+statelength(x) = 0
+
+setup(rng::AbstractRNG, l::AbstractExplicitLayer) = (initialparameters(rng, l), initialstates(rng, l))
+setup(l::AbstractExplicitLayer) = setup(Random.GLOBAL_RNG, l)
+
+apply(model::AbstractExplicitLayer, x, ps::Union{ComponentArray,NamedTuple}, st::NamedTuple) = model(x, ps, st)
```

Comment:
I might be wrong, but it seems they need to pass RNGs since they don't make a distinction between parameters and states? https://flax.readthedocs.io/en/latest/flax.linen.html?highlight=make_rng#flax.linen.Module.apply. In our case, layers should store their own RNG in their state

Reply:
AFAICT they do, but the internals are quite complex so it's hard to figure out where everything is threaded through. I suppose you could include it as an input if need be, but that seems awkward if `x` is not varargs.